File: NetCore/TwoFactorAuth.Core/Services/DeviceManagementService.cs
// SPDX-FileCopyrightText: (C) 2025 NetKnights GmbH <https://netknights.it>
// SPDX-License-Identifier: AGPL-3.0-or-later

using System.Collections.Concurrent;
using Microsoft.Extensions.Logging;
using TwoFactorAuth.Core.Models;
using OtpNet;

namespace TwoFactorAuth.Core.Services;

/// <summary>
/// Interface for device management operations
/// </summary>
public interface IDeviceManagementService
{
    /// <summary>
    /// Request device activation and generate OTP for display
    /// </summary>
    Task<ActivationResponse> RequestDeviceActivationAsync(ActivationRequest request);
    
    /// <summary>
    /// Validate device activation with OTP code
    /// </summary>
    Task<ActivationValidationResult> ValidateDeviceActivationAsync(DeviceActivationValidation validation);
    
    /// <summary>
    /// Check if device is whitelisted and activated
    /// </summary>
    Task<bool> IsDeviceActivatedAsync(string deviceId);
    
    /// <summary>
    /// Get device information
    /// </summary>
    Task<DeviceInfo?> GetDeviceInfoAsync(string deviceId);
    
    /// <summary>
    /// Get all activated devices
    /// </summary>
    Task<IEnumerable<DeviceInfo>> GetActivatedDevicesAsync();
    
    /// <summary>
    /// Deactivate a device
    /// </summary>
    Task<bool> DeactivateDeviceAsync(string deviceId);
}

/// <summary>
/// Device management service implementation
/// </summary>
public class DeviceManagementService : IDeviceManagementService
{
    private readonly ILogger<DeviceManagementService> _logger;
    
    // In-memory storage for demonstration
    // In production, use a database
    private readonly ConcurrentDictionary<string, DeviceInfo> _devices = new();
    private readonly ConcurrentDictionary<string, (string Secret, DateTime ExpiresAt)> _pendingActivations = new(); // deviceId -> (secret, expiry)
    
    // Activation requests expire after 5 minutes
    private const int ActivationExpiryMinutes = 5;
    
    public DeviceManagementService(ILogger<DeviceManagementService> logger)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }
    
    public Task<ActivationResponse> RequestDeviceActivationAsync(ActivationRequest request)
    {
        try
        {
            _logger.LogInformation("Device activation requested for {DeviceId}", request.DeviceId);
            
            // Generate secret for this device
            var secretKey = KeyGeneration.GenerateRandomKey(20);
            var secret = Base32Encoding.ToString(secretKey);
            
            // Generate current OTP code to display
            var totp = new Totp(secretKey, step: 30, totpSize: 6);
            var otpCode = totp.ComputeTotp();
            
            // Store pending activation with expiration
            var expiresAt = DateTime.UtcNow.AddMinutes(ActivationExpiryMinutes);
            _pendingActivations[request.DeviceId] = (secret, expiresAt);
            
            // Store device info as pending
            var deviceInfo = new DeviceInfo
            {
                DeviceId = request.DeviceId,
                DeviceName = request.DeviceName ?? "Unknown Device",
                Platform = request.Platform ?? "Unknown",
                OsVersion = request.OsVersion ?? "Unknown",
                Model = request.Model ?? "Unknown",
                RegisteredAt = DateTime.UtcNow,
                IsActivated = false
            };
            
            _devices.AddOrUpdate(request.DeviceId, deviceInfo, (key, existing) => deviceInfo);
            
            var response = new ActivationResponse
            {
                Success = true,
                Message = "Activation OTP generated. Please enter this code on your device.",
                Secret = secret,
                OtpCode = otpCode,
                DeviceId = request.DeviceId
            };
            
            _logger.LogInformation("Generated activation OTP for device {DeviceId}: {OtpCode}", 
                request.DeviceId, otpCode);
            
            return Task.FromResult(response);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error requesting device activation for {DeviceId}", request.DeviceId);
            return Task.FromResult(new ActivationResponse
            {
                Success = false,
                Message = $"Error requesting activation: {ex.Message}"
            });
        }
    }
    
    public Task<ActivationValidationResult> ValidateDeviceActivationAsync(DeviceActivationValidation validation)
    {
        try
        {
            _logger.LogInformation("Validating device activation for {DeviceId}", validation.DeviceId);
            
            // Check if we have a pending activation for this device
            if (!_pendingActivations.TryGetValue(validation.DeviceId, out var activationData))
            {
                _logger.LogWarning("No pending activation found for device {DeviceId}", validation.DeviceId);
                return Task.FromResult(new ActivationValidationResult
                {
                    Success = false,
                    Message = "No pending activation found for this device. Please request activation first."
                });
            }
            
            // Check if activation has expired
            if (DateTime.UtcNow > activationData.ExpiresAt)
            {
                _pendingActivations.TryRemove(validation.DeviceId, out _);
                _logger.LogWarning("Activation request expired for device {DeviceId}", validation.DeviceId);
                return Task.FromResult(new ActivationValidationResult
                {
                    Success = false,
                    Message = "Activation request has expired. Please request activation again."
                });
            }
            
            var secret = activationData.Secret;
            
            // Validate OTP code
            var secretBytes = Base32Encoding.ToBytes(secret);
            var totp = new Totp(secretBytes, step: 30, totpSize: 6);
            
            long timeStepMatched;
            var isValid = totp.VerifyTotp(validation.OtpCode, out timeStepMatched, 
                new VerificationWindow(1, 1)); // Allow 1 step before and after
            
            if (!isValid)
            {
                _logger.LogWarning("Invalid OTP code for device {DeviceId}", validation.DeviceId);
                return Task.FromResult(new ActivationValidationResult
                {
                    Success = false,
                    Message = "Invalid OTP code. Please try again."
                });
            }
            
            // Mark device as activated
            if (_devices.TryGetValue(validation.DeviceId, out var deviceInfo))
            {
                deviceInfo.IsActivated = true;
                deviceInfo.ActivatedAt = DateTime.UtcNow;
                deviceInfo.ActivationToken = Guid.NewGuid().ToString();
                
                // Remove from pending activations
                _pendingActivations.TryRemove(validation.DeviceId, out _);
                
                _logger.LogInformation("Device {DeviceId} activated successfully", validation.DeviceId);
                
                return Task.FromResult(new ActivationValidationResult
                {
                    Success = true,
                    Message = "Device activated successfully!",
                    ActivationToken = deviceInfo.ActivationToken,
                    ActivatedAt = deviceInfo.ActivatedAt
                });
            }
            else
            {
                _logger.LogError("Device info not found for {DeviceId}", validation.DeviceId);
                return Task.FromResult(new ActivationValidationResult
                {
                    Success = false,
                    Message = "Device information not found."
                });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error validating device activation for {DeviceId}", validation.DeviceId);
            return Task.FromResult(new ActivationValidationResult
            {
                Success = false,
                Message = $"Error validating activation: {ex.Message}"
            });
        }
    }
    
    public Task<bool> IsDeviceActivatedAsync(string deviceId)
    {
        if (_devices.TryGetValue(deviceId, out var deviceInfo))
        {
            return Task.FromResult(deviceInfo.IsActivated);
        }
        return Task.FromResult(false);
    }
    
    public Task<DeviceInfo?> GetDeviceInfoAsync(string deviceId)
    {
        _devices.TryGetValue(deviceId, out var deviceInfo);
        return Task.FromResult(deviceInfo);
    }
    
    public Task<IEnumerable<DeviceInfo>> GetActivatedDevicesAsync()
    {
        var activatedDevices = _devices.Values.Where(d => d.IsActivated).ToList();
        return Task.FromResult<IEnumerable<DeviceInfo>>(activatedDevices);
    }
    
    public Task<bool> DeactivateDeviceAsync(string deviceId)
    {
        if (_devices.TryGetValue(deviceId, out var deviceInfo))
        {
            deviceInfo.IsActivated = false;
            deviceInfo.ActivatedAt = null;
            deviceInfo.ActivationToken = null;
            
            _logger.LogInformation("Device {DeviceId} deactivated", deviceId);
            return Task.FromResult(true);
        }
        
        _logger.LogWarning("Device {DeviceId} not found for deactivation", deviceId);
        return Task.FromResult(false);
    }
}

File: nodejs_package/two-factor-auth/src/services/DeviceManagementService.ts
import { v4 as uuidv4 } from 'uuid';
import { OtpManager } from './OtpManager';
import {
  DeviceInfo,
  ActivationRequest,
  ActivationResponse,
  DeviceActivationValidation,
  ActivationValidationResult,
} from '../models/types';

/**
 * Device Management Service for handling device activation and management
 */
export class DeviceManagementService {
  private readonly otpManager: OtpManager;
  
  // In-memory storage (use database in production)
  private devices: Map<string, DeviceInfo>;
  private pendingActivations: Map<string, { secret: string; expiresAt: Date }>; // deviceId -> { secret, expiresAt }
  
  // Activation requests expire after 5 minutes
  private readonly activationExpiryMinutes = 5;

  constructor(otpManager?: OtpManager) {
    this.otpManager = otpManager || new OtpManager();
    this.devices = new Map();
    this.pendingActivations = new Map();
  }

  /**
   * Request device activation and generate OTP for display
   */
  async requestDeviceActivation(request: ActivationRequest): Promise<ActivationResponse> {
    try {
      console.log(`Device activation requested for ${request.deviceId}`);

      // Generate secret for this device
      const secret = this.otpManager.generateSecret();

      // Generate current OTP code to display (use sync version for simplicity)
      const otpCode = this.otpManager.generateTotpSync(secret);

      // Store pending activation with expiration
      const expiresAt = new Date(Date.now() + this.activationExpiryMinutes * 60 * 1000);
      this.pendingActivations.set(request.deviceId, { secret, expiresAt });

      // Store device info as pending
      const deviceInfo: DeviceInfo = {
        deviceId: request.deviceId,
        deviceName: request.deviceName || 'Unknown Device',
        platform: request.platform || 'Unknown',
        osVersion: request.osVersion || 'Unknown',
        model: request.model || 'Unknown',
        registeredAt: new Date(),
        isActivated: false,
      };

      this.devices.set(request.deviceId, deviceInfo);

      console.log(`Generated activation OTP for device ${request.deviceId}: ${otpCode}`);

      return {
        success: true,
        message: 'Activation OTP generated. Please enter this code on your device.',
        secret,
        otpCode,
        deviceId: request.deviceId,
      };
    } catch (error) {
      console.error('Error requesting device activation:', error);
      return {
        success: false,
        message: `Error requesting activation: ${error instanceof Error ? error.message : 'Unknown error'}`,
      };
    }
  }

  /**
   * Validate device activation with OTP code
   */
  async validateDeviceActivation(
    validation: DeviceActivationValidation
  ): Promise<ActivationValidationResult> {
    try {
      console.log(`Validating device activation for ${validation.deviceId}`);

      // Check if we have a pending activation for this device
      const activationData = this.pendingActivations.get(validation.deviceId);
      if (!activationData) {
        console.warn(`No pending activation found for device ${validation.deviceId}`);
        return {
          success: false,
          message: 'No pending activation found for this device. Please request activation first.',
        };
      }

      // Check if activation has expired
      if (new Date() > activationData.expiresAt) {
        this.pendingActivations.delete(validation.deviceId);
        console.warn(`Activation request expired for device ${validation.deviceId}`);
        return {
          success: false,
          message: 'Activation request has expired. Please request activation again.',
        };
      }

      const secret = activationData.secret;

      // Validate OTP code
      const isValid = await this.otpManager.validateTotp(secret, validation.otpCode);

      if (!isValid) {
        console.warn(`Invalid OTP code for device ${validation.deviceId}`);
        return {
          success: false,
          message: 'Invalid OTP code. Please try again.',
        };
      }

      // Mark device as activated
      const deviceInfo = this.devices.get(validation.deviceId);
      if (deviceInfo) {
        deviceInfo.isActivated = true;
        deviceInfo.activatedAt = new Date();
        deviceInfo.activationToken = uuidv4();

        // Remove from pending activations
        this.pendingActivations.delete(validation.deviceId);

        console.log(`Device ${validation.deviceId} activated successfully`);

        return {
          success: true,
          message: 'Device activated successfully!',
          activationToken: deviceInfo.activationToken,
          activatedAt: deviceInfo.activatedAt,
        };
      } else {
        console.error(`Device info not found for ${validation.deviceId}`);
        return {
          success: false,
          message: 'Device information not found.',
        };
      }
    } catch (error) {
      console.error('Error validating device activation:', error);
      return {
        success: false,
        message: `Error validating activation: ${error instanceof Error ? error.message : 'Unknown error'}`,
      };
    }
  }

  /**
   * Check if device is activated
   */
  async isDeviceActivated(deviceId: string): Promise<boolean> {
    const deviceInfo = this.devices.get(deviceId);
    return deviceInfo?.isActivated || false;
  }

  /**
   * Get device information
   */
  async getDeviceInfo(deviceId: string): Promise<DeviceInfo | null> {
    return this.devices.get(deviceId) || null;
  }

  /**
   * Get all activated devices
   */
  async getActivatedDevices(): Promise<DeviceInfo[]> {
    return Array.from(this.devices.values()).filter((device) => device.isActivated);
  }

  /**
   * Deactivate a device
   */
  async deactivateDevice(deviceId: string): Promise<boolean> {
    const deviceInfo = this.devices.get(deviceId);
    if (deviceInfo) {
      deviceInfo.isActivated = false;
      deviceInfo.activatedAt = undefined;
      deviceInfo.activationToken = undefined;

      console.log(`Device ${deviceId} deactivated`);
      return true;
    }

    console.warn(`Device ${deviceId} not found for deactivation`);
    return false;
  }

  /**
   * Get all devices (for admin purposes)
   */
  async getAllDevices(): Promise<DeviceInfo[]> {
    return Array.from(this.devices.values());
  }

  /**
   * Clear all devices (for testing purposes)
   */
  async clearAllDevices(): Promise<void> {
    this.devices.clear();
    this.pendingActivations.clear();
  }
}
